File: integration-tests/cpp/coherent-sets/src/subscriber.cpp
/*
 * integration-tests/cpp/coherent-sets -- subscriber. Direct C++ port of
 * c/coherent-sets/src/subscriber.c -- see that file's header comment for
 * the full atomicity-assertion rationale and
 * docs/design/integration-test-tier.md for the scenario spec.
 *
 * Required stdout markers: "Create topic:" x2, "Create reader for topic:"
 * x2, "Subscriber: group N paired (position+velocity).", "Subscriber:
 * received all 20 groups, atomic and ordered." Any failure path prints a
 * line starting "FAIL:" and exits nonzero.
 */
#include "pose_group.hpp"
#include "zzdds_cpp.hpp"
#include "dcps_impl.hpp"

#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <memory>
#include <vector>

namespace {

constexpr int GROUP_COUNT = 20;
constexpr ::DDS::Duration_t WAIT_STEP{1, 0};
constexpr int OVERALL_DEADLINE_MS = 30000;

uint32_t parse_domain(int argc, char **argv) {
    for (int i = 1; i < argc - 1; i++) {
        if (std::strcmp(argv[i], "-d") == 0 || std::strcmp(argv[i], "--domain") == 0) {
            return static_cast<uint32_t>(std::strtoul(argv[i + 1], nullptr, 10));
        }
    }
    return 0;
}

} // namespace

int main(int argc, char **argv) {
    uint32_t domain_id = parse_domain(argc, argv);

    auto factory = zzdds::create_factory();
    if (!factory) {
        std::fprintf(stderr, "FAIL: createFactory() failed\n");
        return 1;
    }

    auto dp = factory->create_participant(domain_id, ::DDS::DomainParticipantQos::default_value(), nullptr, 0);
    if (!dp) {
        std::fprintf(stderr, "FAIL: create_participant() failed on domain %u\n", domain_id);
        return 1;
    }
    auto dp_handle = dp->native_handle();

    if (PositionTypeSupport::register_type(dp_handle) != 0) {
        std::fprintf(stderr, "FAIL: register PositionTypeSupport failed\n");
        return 1;
    }
    if (VelocityTypeSupport::register_type(dp_handle) != 0) {
        std::fprintf(stderr, "FAIL: register VelocityTypeSupport failed\n");
        return 1;
    }

    auto position_topic = dp->create_topic("Position", "Position", ::DDS::TopicQos::default_value(), nullptr, 0);
    if (!position_topic) {
        std::fprintf(stderr, "FAIL: create_topic(Position) failed\n");
        return 1;
    }
    std::printf("Create topic: Position\n");

    auto velocity_topic = dp->create_topic("Velocity", "Velocity", ::DDS::TopicQos::default_value(), nullptr, 0);
    if (!velocity_topic) {
        std::fprintf(stderr, "FAIL: create_topic(Velocity) failed\n");
        return 1;
    }
    std::printf("Create topic: Velocity\n");

    auto sub_qos = ::DDS::SubscriberQos::default_value();
    sub_qos.presentation.access_scope = ::DDS::PresentationQosPolicyAccessScopeKind::GROUP_PRESENTATION_QOS;
    sub_qos.presentation.coherent_access = true;
    sub_qos.presentation.ordered_access = true;

    auto sub = dp->create_subscriber(sub_qos, nullptr, 0);
    if (!sub) {
        std::fprintf(stderr, "FAIL: create_subscriber() failed\n");
        return 1;
    }

    auto dr_qos = ::DDS::DataReaderQos::default_value();
    dr_qos.reliability.kind = ::DDS::ReliabilityQosPolicyKind::RELIABLE_RELIABILITY_QOS;
    dr_qos.history.kind = ::DDS::HistoryQosPolicyKind::KEEP_ALL_HISTORY_QOS;

    // See cpp/waitset/src/subscriber.cpp's matching comment: create_topic()
    // constructs zzdds::detail::TopicSupport under the hood, so this upcast
    // is real.
    auto zposition_topic = std::static_pointer_cast<::zzdds::TopicImpl>(position_topic);
    auto position_desc = zposition_topic->as_topic_description();
    auto position_dr = sub->create_datareader(position_desc, dr_qos, nullptr, 0);
    if (!position_dr) {
        std::fprintf(stderr, "FAIL: create_datareader(Position) failed\n");
        return 1;
    }
    std::printf("Create reader for topic: Position\n");

    auto zvelocity_topic = std::static_pointer_cast<::zzdds::TopicImpl>(velocity_topic);
    auto velocity_desc = zvelocity_topic->as_topic_description();
    auto velocity_dr = sub->create_datareader(velocity_desc, dr_qos, nullptr, 0);
    if (!velocity_dr) {
        std::fprintf(stderr, "FAIL: create_datareader(Velocity) failed\n");
        return 1;
    }
    std::printf("Create reader for topic: Velocity\n");

    PositionDataReader position_reader(position_dr->native_handle());
    VelocityDataReader velocity_reader(velocity_dr->native_handle());

    auto ws = zzdds::create_waitset();
    if (!ws) {
        std::fprintf(stderr, "FAIL: create_waitset() failed\n");
        return 1;
    }

    auto position_rc = position_dr->create_readcondition(::DDS::ANY_SAMPLE_STATE, ::DDS::ANY_VIEW_STATE, ::DDS::ANY_INSTANCE_STATE);
    auto velocity_rc = velocity_dr->create_readcondition(::DDS::ANY_SAMPLE_STATE, ::DDS::ANY_VIEW_STATE, ::DDS::ANY_INSTANCE_STATE);
    if (!position_rc || !velocity_rc) {
        std::fprintf(stderr, "FAIL: create_readcondition() failed\n");
        return 1;
    }
    if (ws->attach_condition(position_rc) != ::DDS::RETCODE_OK ||
        ws->attach_condition(velocity_rc) != ::DDS::RETCODE_OK)
    {
        std::fprintf(stderr, "FAIL: attach_condition() failed\n");
        return 1;
    }

    std::vector<int> position_order, velocity_order;
    position_order.reserve(GROUP_COUNT);
    velocity_order.reserve(GROUP_COUNT);

    int overall_waited_ms = 0;
    while (static_cast<int>(position_order.size()) < GROUP_COUNT || static_cast<int>(velocity_order.size()) < GROUP_COUNT) {
        if (overall_waited_ms >= OVERALL_DEADLINE_MS) {
            std::fprintf(stderr, "FAIL: only received position=%zu velocity=%zu/%d within %ds\n",
                          position_order.size(), velocity_order.size(), GROUP_COUNT, OVERALL_DEADLINE_MS / 1000);
            return 1;
        }

        ::DDS::ConditionSeq active;
        auto wr = ws->wait(active, WAIT_STEP);
        if (wr == ::DDS::RETCODE_TIMEOUT) {
            overall_waited_ms += 1000;
            continue;
        }
        if (wr != ::DDS::RETCODE_OK) {
            std::fprintf(stderr, "FAIL: WaitSet.wait() returned %d\n", static_cast<int>(wr));
            return 1;
        }

        if (sub->begin_access() != ::DDS::RETCODE_OK) {
            std::fprintf(stderr, "FAIL: begin_access() failed\n");
            return 1;
        }

        std::vector<::Position> position_values(GROUP_COUNT);
        std::vector<DDS_SampleInfo> position_infos(GROUP_COUNT);
        int n_pos_taken = position_reader.take_n(position_values.data(), position_infos.data(), GROUP_COUNT,
                                                  ::DDS::ANY_SAMPLE_STATE, ::DDS::ANY_VIEW_STATE, ::DDS::ANY_INSTANCE_STATE);
        for (int i = 0; i < n_pos_taken; i++) {
            if (!position_infos[i].valid_data) continue;
            position_order.push_back(position_values[i].group_id);
        }

        std::vector<::Velocity> velocity_values(GROUP_COUNT);
        std::vector<DDS_SampleInfo> velocity_infos(GROUP_COUNT);
        int n_vel_taken = velocity_reader.take_n(velocity_values.data(), velocity_infos.data(), GROUP_COUNT,
                                                  ::DDS::ANY_SAMPLE_STATE, ::DDS::ANY_VIEW_STATE, ::DDS::ANY_INSTANCE_STATE);
        for (int i = 0; i < n_vel_taken; i++) {
            if (!velocity_infos[i].valid_data) continue;
            velocity_order.push_back(velocity_values[i].group_id);
        }

        if (sub->end_access() != ::DDS::RETCODE_OK) {
            std::fprintf(stderr, "FAIL: end_access() failed\n");
            return 1;
        }

        // The core atomicity assertion -- see c/coherent-sets/src/subscriber.c's header comment.
        if (position_order.size() != velocity_order.size()) {
            std::fprintf(stderr,
                          "FAIL: atomicity violated -- position and velocity readers diverged after an access bracket "
                          "(position count=%zu velocity count=%zu) -- a group became visible on one reader without its pair\n",
                          position_order.size(), velocity_order.size());
            return 1;
        }
        for (size_t i = 0; i < position_order.size(); i++) {
            if (position_order[i] != velocity_order[i]) {
                std::fprintf(stderr, "FAIL: atomicity violated at index %zu -- position group_id=%d but velocity group_id=%d\n",
                              i, position_order[i], velocity_order[i]);
                return 1;
            }
        }
        if (!position_order.empty()) {
            std::printf("Subscriber: group %d paired (position+velocity).\n", position_order.back());
        }
    }

    for (int i = 0; i < GROUP_COUNT; i++) {
        if (position_order[i] != i || velocity_order[i] != i) {
            std::fprintf(stderr, "FAIL: ordered_access violated at index %d -- expected group_id=%d, got position=%d velocity=%d\n",
                          i, i, position_order[i], velocity_order[i]);
            return 1;
        }
    }

    std::printf("Subscriber: received all %d groups, atomic and ordered.\n", GROUP_COUNT);

    ws->detach_condition(position_rc);
    ws->detach_condition(velocity_rc);
    position_dr->delete_readcondition(position_rc);
    velocity_dr->delete_readcondition(velocity_rc);
    sub->delete_datareader(position_dr);
    sub->delete_datareader(velocity_dr);

    return 0;
}

File: integration-tests/java/coherent-sets/Subscriber.java
// integration-tests/java/coherent-sets -- subscriber. Direct Java port of
// c/coherent-sets/src/subscriber.c -- see that file's header comment for
// the full atomicity-assertion rationale and
// docs/design/integration-test-tier.md for the scenario spec.
//
// Required stdout markers: "Create topic:" x2, "Create reader for topic:"
// x2, "Subscriber: group N paired (position+velocity).", "Subscriber:
// received all 20 groups, atomic and ordered." Any failure path prints a
// line starting "FAIL:" and exits nonzero.

import io.zzdds.dcps.Dcps;

import java.util.ArrayList;
import java.util.List;

public class Subscriber {
    static final int GROUP_COUNT = 20;
    static final int WAIT_STEP_SEC = 1;
    static final int OVERALL_DEADLINE_MS = 30000;

    static int parseDomain(String[] args) {
        for (int i = 0; i < args.length - 1; i++) {
            if (args[i].equals("-d") || args[i].equals("--domain")) {
                return Integer.parseInt(args[i + 1]);
            }
        }
        return 0;
    }

    public static void main(String[] args) throws Exception {
        int domainId = parseDomain(args);

        Dcps.DDS.DomainParticipantFactory factory =
            (Dcps.DDS.DomainParticipantFactory) io.zzdds.runtime.ZzddsRuntime.createFactory();
        if (factory == null) {
            System.err.println("FAIL: createFactory() failed");
            System.exit(1);
        }

        Dcps.DDS.DomainParticipant dp = factory.create_participant(domainId, null, null, 0);
        if (dp == null) {
            System.err.println("FAIL: create_participant() failed on domain " + domainId);
            System.exit(1);
        }

        if (PositionTypeSupport.register(dp, "Position") != 0) {
            System.err.println("FAIL: register PositionTypeSupport failed");
            System.exit(1);
        }
        if (VelocityTypeSupport.register(dp, "Velocity") != 0) {
            System.err.println("FAIL: register VelocityTypeSupport failed");
            System.exit(1);
        }

        Dcps.DDS.Topic positionTopic = dp.create_topic("Position", "Position", null, null, 0);
        if (positionTopic == null) {
            System.err.println("FAIL: create_topic(Position) failed");
            System.exit(1);
        }
        System.out.println("Create topic: Position");

        Dcps.DDS.Topic velocityTopic = dp.create_topic("Velocity", "Velocity", null, null, 0);
        if (velocityTopic == null) {
            System.err.println("FAIL: create_topic(Velocity) failed");
            System.exit(1);
        }
        System.out.println("Create topic: Velocity");

        Dcps.DDS.SubscriberQos subQos = new Dcps.DDS.SubscriberQos();
        subQos.get_presentation().set_access_scope(Dcps.DDS.PresentationQosPolicyAccessScopeKind.GROUP_PRESENTATION_QOS);
        subQos.get_presentation().set_coherent_access(true);
        subQos.get_presentation().set_ordered_access(true);

        Dcps.DDS.Subscriber sub = dp.create_subscriber(subQos, null, 0);
        if (sub == null) {
            System.err.println("FAIL: create_subscriber() failed");
            System.exit(1);
        }

        Dcps.DDS.DataReaderQos drQos = new Dcps.DDS.DataReaderQos();
        sub.get_default_datareader_qos(drQos);
        drQos.get_reliability().set_kind(Dcps.DDS.ReliabilityQosPolicyKind.RELIABLE_RELIABILITY_QOS);
        drQos.get_history().set_kind(Dcps.DDS.HistoryQosPolicyKind.KEEP_ALL_HISTORY_QOS);

        Dcps.DDS.DataReader positionDr = sub.create_datareader(positionTopic, drQos, null, 0);
        if (positionDr == null) {
            System.err.println("FAIL: create_datareader(Position) failed");
            System.exit(1);
        }
        System.out.println("Create reader for topic: Position");

        Dcps.DDS.DataReader velocityDr = sub.create_datareader(velocityTopic, drQos, null, 0);
        if (velocityDr == null) {
            System.err.println("FAIL: create_datareader(Velocity) failed");
            System.exit(1);
        }
        System.out.println("Create reader for topic: Velocity");

        PositionDataReader positionReader = new PositionDataReader(positionDr);
        VelocityDataReader velocityReader = new VelocityDataReader(velocityDr);

        Dcps.DDS.WaitSet ws = (Dcps.DDS.WaitSet) io.zzdds.runtime.ZzddsRuntime.createWaitSet();
        if (ws == null) {
            System.err.println("FAIL: createWaitSet() failed");
            System.exit(1);
        }

        Dcps.DDS.ReadCondition positionRc = positionDr.create_readcondition(
            Dcps.DDS.ANY_SAMPLE_STATE.value, Dcps.DDS.ANY_VIEW_STATE.value, Dcps.DDS.ANY_INSTANCE_STATE.value);
        Dcps.DDS.ReadCondition velocityRc = velocityDr.create_readcondition(
            Dcps.DDS.ANY_SAMPLE_STATE.value, Dcps.DDS.ANY_VIEW_STATE.value, Dcps.DDS.ANY_INSTANCE_STATE.value);
        if (positionRc == null || velocityRc == null) {
            System.err.println("FAIL: create_readcondition() failed");
            System.exit(1);
        }
        if (ws.attach_condition(positionRc) != 0 || ws.attach_condition(velocityRc) != 0) {
            System.err.println("FAIL: attach_condition() failed");
            System.exit(1);
        }

        List<Integer> positionOrder = new ArrayList<>();
        List<Integer> velocityOrder = new ArrayList<>();

        Dcps.DDS.Duration_t waitStep = new Dcps.DDS.Duration_t();
        waitStep.set_sec(WAIT_STEP_SEC);
        waitStep.set_nanosec(0);

        int overallWaitedMs = 0;
        while (positionOrder.size() < GROUP_COUNT || velocityOrder.size() < GROUP_COUNT) {
            if (overallWaitedMs >= OVERALL_DEADLINE_MS) {
                System.err.println("FAIL: only received position=" + positionOrder.size() + " velocity=" + velocityOrder.size()
                    + "/" + GROUP_COUNT + " within " + (OVERALL_DEADLINE_MS / 1000) + "s");
                System.exit(1);
            }

            ArrayList<Dcps.DDS.Condition> active = new ArrayList<>();
            int wr = ws.wait(active, waitStep);
            if (wr == Dcps.DDS.RETCODE_TIMEOUT.value) {
                overallWaitedMs += WAIT_STEP_SEC * 1000;
                continue;
            }
            if (wr != 0) {
                System.err.println("FAIL: WaitSet.wait() returned " + wr);
                System.exit(1);
            }

            if (sub.begin_access() != 0) {
                System.err.println("FAIL: begin_access() failed");
                System.exit(1);
            }

            PositionDataReader.Sample[] positionSamples = positionReader.take_n(
                GROUP_COUNT, Dcps.DDS.ANY_SAMPLE_STATE.value, Dcps.DDS.ANY_VIEW_STATE.value, Dcps.DDS.ANY_INSTANCE_STATE.value);
            for (PositionDataReader.Sample s : positionSamples) {
                if (!s.validData) continue;
                positionOrder.add(s.data.get_group_id());
            }

            VelocityDataReader.Sample[] velocitySamples = velocityReader.take_n(
                GROUP_COUNT, Dcps.DDS.ANY_SAMPLE_STATE.value, Dcps.DDS.ANY_VIEW_STATE.value, Dcps.DDS.ANY_INSTANCE_STATE.value);
            for (VelocityDataReader.Sample s : velocitySamples) {
                if (!s.validData) continue;
                velocityOrder.add(s.data.get_group_id());
            }

            if (sub.end_access() != 0) {
                System.err.println("FAIL: end_access() failed");
                System.exit(1);
            }

            // The core atomicity assertion -- see c/coherent-sets/src/subscriber.c's header comment.
            if (positionOrder.size() != velocityOrder.size()) {
                System.err.println("FAIL: atomicity violated -- position and velocity readers diverged after an access bracket "
                    + "(position count=" + positionOrder.size() + " velocity count=" + velocityOrder.size()
                    + ") -- a group became visible on one reader without its pair");
                System.exit(1);
            }
            for (int i = 0; i < positionOrder.size(); i++) {
                if (!positionOrder.get(i).equals(velocityOrder.get(i))) {
                    System.err.println("FAIL: atomicity violated at index " + i + " -- position group_id=" + positionOrder.get(i)
                        + " but velocity group_id=" + velocityOrder.get(i));
                    System.exit(1);
                }
            }
            if (!positionOrder.isEmpty()) {
                System.out.println("Subscriber: group " + positionOrder.get(positionOrder.size() - 1) + " paired (position+velocity).");
            }
        }

        for (int i = 0; i < GROUP_COUNT; i++) {
            if (positionOrder.get(i) != i || velocityOrder.get(i) != i) {
                System.err.println("FAIL: ordered_access violated at index " + i + " -- expected group_id=" + i
                    + ", got position=" + positionOrder.get(i) + " velocity=" + velocityOrder.get(i));
                System.exit(1);
            }
        }

        System.out.println("Subscriber: received all " + GROUP_COUNT + " groups, atomic and ordered.");

        ws.detach_condition(positionRc);
        ws.detach_condition(velocityRc);
        positionDr.delete_readcondition(positionRc);
        velocityDr.delete_readcondition(velocityRc);
        sub.delete_datareader(positionDr);
        sub.delete_datareader(velocityDr);
    }
}
